### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by johnwhitton**
Refactor to improve solidity tooling using hardhat including testing, deployment, documentation and gas estimation. As well as consolidate all tools, scripts and cli into a src folder.

### Migration Strategy
* Smart Contract use Hardhat with Typescript and ethers(instead of web3)
  * Replace all web3 with ethers
  * replace all solidity deployments and test js files with typescript
  * remove all truffleConfig use hardhat instead
  * write tests
* docs: new folder for documentation
* docs/assets => migrated from assets
* docs/solidity: contains generated solidity documentation
* deploy: new folder for deployment scripts (using hardhat-deploy and logic from scripts)
* src: new folder for typescript source files 
* src/lib: (migrated from scripts)
* src/cli: (migrated from cli)
* src/(elc, eprover, eth2hmy-relay): migrated from tools(elc, eprover, eth2hmy-relay)

### Action Items
- [ ] Resolve gas issue with Kovan Deployment
- [ ] Resolve upgrade Process
- [ ] Generate DAGs for Local Ethereum (Hardhat) node

### Logs and Analysis

```
johnlaptop horizon (refactor) $ npx hardhat deploy --network hardhat --tags HarmonyLightClient
Nothing to compile
No need to generate any newer typings.
✅ Generated documentation for 46 contracts
(node:62043) UnhandledPromiseRejectionWarning: Error: processing response error (body="{\"jsonrpc\":\"2.0\",\"id\":42,\"error\":{\"code\":-32600,\"message\":\"invalid request\"}}\n", error={"code":-32600}, requestBody="{\"method\":{\"jsonrpc\":\"2.0\",\"method\":\"hmyv2_getFullHeader\",\"params\":[\"0xe\"],\"id\":1654537992451},\"id\":42,\"jsonrpc\":\"2.0\"}", requestMethod="POST", url="http://localhost:9500", code=SERVER_ERROR, version=web/5.6.0)
    at Logger.makeError (/Users/john/one-wallet/horizon/node_modules/@ethersproject/logger/src.ts/index.ts:261:28)
    at Logger.throwError (/Users/john/one-wallet/horizon/node_modules/@ethersproject/logger/src.ts/index.ts:273:20)
    at /Users/john/one-wallet/horizon/node_modules/@ethersproject/web/src.ts/index.ts:329:28
    at step (/Users/john/one-wallet/horizon/node_modules/@ethersproject/web/lib/index.js:33:23)
    at Object.next (/Users/john/one-wallet/horizon/node_modules/@ethersproject/web/lib/index.js:14:53)
    at fulfilled (/Users/john/one-wallet/horizon/node_modules/@ethersproject/web/lib/index.js:5:58)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:62043) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:62043) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -16,3 +16,17 @@ cache
...
+#Hardhat files
+cache
+artifacts
+
+# OS metadata
+.DS_Store
+Thumbs.db
+
+# Ignore built ts files
+dist/**/*
```
Comment (polymorpher) [inline .gitignore:32]: This will ignore eth2hmy-relay/ethash/dist/**

File: docs/README.md [hunk 2]
```diff
@@ -1,2 +1,294 @@
 # Horizon Bridge Documentation and API reference
 
+# Overview
```
Comment (polymorpher) [inline docs/README.md:3]: Was this updated after the other PR (#37)?

File: docs/SOLIDITY.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
+# Advanced Sample Hardhat Project
```
Comment (polymorpher) [inline docs/SOLIDITY.md:1]: Remove template file? Doesn't seem to be very relevant

Comment (polymorpher) [reply docs/SOLIDITY.md:1]: If some commands are helpful - we can keep them in a short cheatsheet

**Comment by polymorpher** [comment]
Can data/abi/* be removed as well? They don't look like abi and they are not referenced elsewhere

**Comment by johnwhitton** [comment]
Merged latest upstream/main still have some action items for testing and deployment which can be progressed separately.
- [ ] confirm linting has an indentation of 4 spaces
- [ ] Fix test `EthLightClient.js`
- [ ] Fix deploy script `004_deploy_ethereumLightClient.ts`
- [ ] Remove scripts directory (util.js -> src/lib and newtest.js -> test)

**Review by gupadhyaya — APPROVED** [review]

**Pull request merged**